File: EDF/EDF_deidentify.m
function EDF_deidentify(filename, copy_file)
%Check to see if any file has been given
if nargin == 0
    [filename, path] = uigetfile('*.*',...
        'Select EDF File(s)', ...
        'MultiSelect', 'on');
    answer = questdlg('Do you want to save a copy or overwrite?', ...
        'Deidentification Menu', ...
        'Save a Copy', 'Overwrite', 'Cancel', 'Cancel');
    
    switch answer
        case 'Save a Copy'
            copy_file = true;
        case 'Overwrite'
            answer2 = questdlg('Are you sure you want to overwrite? Original information will be lost.', ...
                'Deidentification Menu', ...
                'Yes (Overwrite)', 'Cancel', 'Cancel');
            
            if strcmpi(answer2,'Yes (Overwrite)')
                copy_file = false;
            else
                return;
            end
            
        otherwise
            return;
    end
    
else
    path = ''; %Blank path
end

%If a single file, wrap in cell to work with the loop
if ~iscell(filename)
    filename = {filename};
end

%Loop through all files and deidentify
for ii = 1:length(filename)
    filename_full = fullfile(path, filename{ii});
    
    if ~exist(filename_full,'file')
        error(['Invalid filename: ' filename_full]);
    end
    
    if copy_file
        filename_new = [filename_full(1:end-4) '_deidentified.edf'];
        copyfile(filename_full, filename_new);
        
        deidentify(filename_new);
    else
        deidentify(filename_full);
    end
end

function deidentify(filename)
%Open the file to read/write without resetting it to an empty file
fid = fopen(filename, 'r+');

fprintf(fid, '%-8s', '0');   % Version must be 0
fprintf(fid, '%-80s', 'DEIDENTIFIED'); % Remove patient info
fprintf(fid,'%-80s', 'DEIDENTIFIED'); %Remove recording info
fprintf(fid, '%02i.%02i.%02i', 1,1,1); % Set date as 01.01.01

fclose(fid);

File: EDF/EDFdeidentify.py
# This software is distributed under the MIT License: https://mit-license.org/

import os
import sys
import shutil
from tkinter import *
from tkinter import filedialog
from tkinter import messagebox

# Load the file(s)
root = Tk()
root.geometry('400x100')
root.title('EDF Deidentifier')

# Put some text on the blank window
canvas = Canvas(root, width=400, height=100, bg="white")
canvas.pack()
canvas.create_text(200, 50, text="EDF Deidentifier", font="Times 30 bold",)

root.withdraw()

# Load the file
root.filenames = filedialog.askopenfilenames(initialdir="~/", title="Select EDF file(s)",
                                             filetypes=(("EDF files", "*.edf"), ("All files", "*.*")))
# Handle file cancel
if root.filenames == '':
    root.destroy()
    sys.exit("Closing executable...")

# Check if user wants to overwrite data
copy_file = messagebox.askyesnocancel(title="EDF Deidentifier", message="Would you like to save the a copy of data?")

# Exit if cancel
if copy_file is None:
    root.destroy()
    sys.exit("Closing executable...")

# Double check!
if not copy_file:
    result = messagebox.askyesno(title="EDF Deidentifier", message="Are you sure you want to overwrite all files?")
    if not result:
        root.destroy()
        sys.exit("Closing executable...")

# Loop through all file names
for path in root.filenames:

    # Skip if bad file
    if not (os.path.isfile(path)):
        print("Invalid file: " + path)
        continue

    # Copy file to new name
    if copy_file:
        file_savedir = filedialog.askdirectory()
        path_new = file_savedir + '/' + os.path.basename(path)[0:-4] + '_deidentified.edf'
        shutil.copy(path, path_new)
        path = path_new

    # Open file(s) and deidentify
    f = open(path, "r+", encoding="iso-8859-1")  # 'r' = read
    try:
        f.write('%-8s' % "0")
        f.write('%-80s' % "DEIDENTIFIED")  # Remove patient info
        f.write('%-80s' % "DEIDENTIFIED")  # Remove recording info
        f.write('01.01.01')  # Set date as 01.01.01
    except UnicodeDecodeError:
        f.close()
        f = open(path, "r+", encoding="iso-8859-2")  # 'r' = read
        try:
            f.write('%-8s' % "0")
            f.write('%-80s' % "DEIDENTIFIED")  # Remove patient info
            f.write('%-80s' % "DEIDENTIFIED")  # Remove recording info
            f.write('01.01.01')  # Set date as 01.01.01
        except UnicodeDecodeError:
            f.close()
            f = open(path, "r+", encoding="utf-8")  # 'r' = read
            try:
                f.write('%-8s' % "0")
                f.write('%-80s' % "DEIDENTIFIED")  # Remove patient info
                f.write('%-80s' % "DEIDENTIFIED")  # Remove recording info
                f.write('01.01.01')  # Set date as 01.01.01
                f.close()
            finally:
                print('No valid encoding format found')

    root.destroy()

